scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala

```diff
@@ -383,52 +385,183 @@ class FormatWriter(formatOps: FormatOps) {
-              val spaces = getIndentation(prevState.indentation + extraIndent)
-              Some(matcher.replaceFirst(s"/**\n$spaces* $$1\n$spaces*/"))
-            case _ => None
+          matcher.find() && {
+            style.docstrings.oneline match {
+              case Docstrings.Oneline.fold =>
+                sb.append("/** ").append(matcher.group(1)).append(" */")
+                true
+              case Docstrings.Oneline.unfold =>
+                val extraIndent = if (style.docstrings.isScalaDoc) 2 else 1
+                val spaces = getIndentation(prevState.indentation + extraIndent)
+                sb.append("/**\n").append(spaces).append("* ")
+                sb.append(matcher.group(1))
+                sb.append('\n').append(spaces).append("*/")
+                true
+              case _ => false
+            }
           }
-          res.filter(_ != text)
         }
       }
 
-      private def formatDocstring(text: String): String = {
+      private def formatDocstring(
+          text: String
+      )(implicit sb: StringBuilder): Unit = {
+        if (style.docstrings.style.isEmpty) sb.append(text)
+        else if (!formatOnelineDocstring(text)) formatMultilineDocstring(text)
+      }
+
+      private def formatMultilineDocstring(
+          text: String
+      )(implicit sb: StringBuilder): Unit = {
         val spaces: String = getIndentation(
           prevState.indentation + (if (style.docstrings.isScalaDoc) 2 else 1)
         )
-        leadingAsteriskSpace.matcher(text).replaceAll(spaces)
+        sb.append(leadingAsteriskSpace.matcher(text).replaceAll(spaces))
       }
 
-      private def formatMultilineComment(text: String): String = {
-        val spaces: String = getIndentation(prevState.indentation + 1)
-        leadingAsteriskSpace.matcher(text).replaceAll(spaces)
-      }
+      private class FormatComment(text: String)(implicit sb: StringBuilder) {
+        private val breakBefore = curr.hasBreakBefore
+        private val indent =
+          if (breakBefore) prevState.indentation
+          else prevState.prev.indentation
+        private val spaces: String = getIndentation(indent + 1)
+        // 2 is for "/*" or " *"
+        private val maxLength = style.maxColumn - indent - 2
+
+        private def getFirstLineLength =
+          if (curr.hasBreakBefore) 0
+          else
+            prevState.prev.column - indent +
+              prevState.split.modification.length
+
+        private def canRewrite =
+          style.comments.wrap match {
+            case Comments.Wrap.no => false
+            case Comments.Wrap.trailing => curr.hasBreakAfter
+            case Comments.Wrap.standalone => breakBefore && curr.hasBreakAfter
+          }
+
+        def formatMultilineComment: Unit = {
+          // don't rewrite comments which contain nested comments
+          if (canRewrite && text.lastIndexOf("/*") == 0) {
+            val sectionIter = new Iterator[Iterator[Iterator[String]]] {
```

Comment:
I don't sure should be type aliases helpful here but just 3-level iterator looks little bit scary

```scala
type WordsIterator = Iterator[String]
type ParagraphIterator = Iterator[WordsIterator]
type SectionIterator = Iterator[ParagraphIterator]
```